Pause management

frnzee opened the pull request:
- added pause menu
- added listener for pause button (esc)
- added pause manager

Work in progress, pause menu still not showing up despite it appears on the scene

frnzee committed:
[WIP] Pause management

- added pause menu prefab
- added listener for pause button (esc)
- added pause manager

ega-ega commented on Assets/Prefabs/PauseMenu.prefab:
This prefab mu be located in UI/Popups/

Pls move to this folder

frnzee replied:
This one was duplicate, removed

ega-ega commented on Assets/Scripts/Installer/GameInstaller.cs:
```diff
@@ -82,5 +85,22 @@ private IGameRun GetGameRun()
             return null;
         }
+
+        private void BindPauseMenuObserver()
```
Combine methods BindPauseMenuObserver and BindPauseManager to  one BindPauseManager

frnzee replied:
Done

ega-ega commented on Assets/Scripts/Services/Pause/IPause.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public interface IPause
+    {
+        public bool Paused
```
remove word public, by default all interface members is public. 
And pls edit this property to "bool Paused { get; set; }" (one line )

frnzee replied:
Done

ega-ega commented on Assets/Scripts/Services/Pause/IPause.cs:
```diff
@@ -0,0 +1,16 @@
+        }
+        
+        event Action<IPause> PauseChanged;
+
```
remove unnecessary line ))

frnzee replied:
Done

ega-ega commented on Assets/Scripts/UI/Views/Popups/Pause/PauseMenu.cs:
```diff
@@ -0,0 +1,28 @@
+            _closeButton.onClick.RemoveListener(Pause);
+        }
+
+        private void Pause()
```
Rename method to "OnCloseButtonClicked"

We use for handlers naming : "On{event name}"

frnzee replied:
Done

ega-ega commented on Assets/Scripts/UI/Views/Popups/Pause/PauseMenu.cs:
```diff
@@ -0,0 +1,28 @@
+
+        private void Pause()
+        {
+            _pauseManager.Paused = !_pauseManager.Paused;
```
Pls set "false" manualy. Because its correct behaviour for close pause menu.
We can get error behaviour if _pauseManager.Paused be in False state

frnzee replied:
Done

ega-ega commented on Assets/Scripts/UI/Views/Popups/Pause/PauseMenuObserver.cs:
```diff
@@ -0,0 +1,35 @@
+            {
+                if (_pauseMenu)
+                {
+                    Debug.Log("Pause menu observer: Pause menu closed");
```
Remove debug

ega-ega replied:
We leave the output of debug information only if it is necessary for logging errors or important game states (scene loading, player death).

frnzee replied:
Done

ega-ega commented on Assets/Scripts/UI/Views/Popups/Pause/PauseMenuObserver.cs:
```diff
@@ -0,0 +1,35 @@
+                else
+                {
+                    _pauseMenu = _uiRoot.Open<PauseMenu>();
+                    Debug.Log("Pause menu observer: Pause menu opened");
```
Remove debug

frnzee replied:
Done

frnzee committed:
Refactoring

- removed canvas components from pause menu prefab to make it showing up in parent UIRoot object
- fixed behaviour of handling esc button and close menu button
- removed duplicate pause menu prefab

ega-ega commented on Assets/Prefabs/PauseObserver.prefab:
This is dublicate? Pls remove

ega-ega commented on Assets/Scripts/Services/Pause/IPause.cs:
```diff
@@ -0,0 +1,7 @@
+    public interface IPause
+    {
+        bool Paused { get; set; }
+    }
```
Pls add event Action<IPause> PauseChanged;

frnzee committed:
Refactoring

- removed duplicate pause observer prefab
- switched back pause changed event to IPause interface as was intended

frnzee committed:
Refactoring

- removed unnecessary public declaration of event in IPause interface
- extracted checking of pause menu existence to a new method
